integrate/simpsons.go

```diff
@@ -0,0 +1,124 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"math"
+	"sort"
+)
+
+// Simpson's method approximates the integral of a function f(x)
+// by means of subdividing the interval of integration into segments
+// and applying a method that fits a polynomial to each subinterval.
+// this implementation makes use of the following composite Simpson's method
+// to estimate \int_a^b f(x) dx:
+//
+// \sum_{i=0}^{(N/2)-1} {a_0}_i * f_{2i+1} + {a_1}_i * f_{2i+1} + {a_2}_i * f_{2i}
+//
+// where N is the number of subintervals and the number of subintervals
+// is even and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
+//
+// {a_0}_i = 2 * h^{3}_{2i+1} - h^{3}_{2i} + 3 * h_{2i} * h^{2}_{2i+1} /
+//           6 * h_{2i+1} * (h_{2i+1} + h_{2i})
+//
+// {a_1}_i = h^{3}_{2i+1} + h^{3}_{2i} +
+//           3 * h_{2i+1} * (h_{2i+1} + h_{2i}) /
+//           6 * h_{2i+1} * h_{2i}
+//
+// {a_2}_i = 2 * h^{3}_{2i} - h^{3}_{2i+1} + 3 * h_{2i+1} * h^{2}_{2i} /
+//           6 * h_{2i} * (h_{2i+1} + h_{2i})
+//
+// The formula above approximates the integral of function of f
+// if N is an even number of subintervals.
+//
+// If the number of subintervals are odd,
+// the subintervals up i=0..n-2 are given by the above
+// and the approximations over the second to last and last subintervals
+// are given by:
+//
+// {a_0}_i * f_{N} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N-1}
+//
+// where the coefficients are:
+//
+// {a_0}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
+//           6 * (h_{N-2} + h_{N-1})
+//
+// {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
+//           6 * h_{N-2}
+//
+// {a_2}_i = -1 * h^{3}_{N-1} /
+//            6 * h_{N-2} * (h_{N-2} + h_{N-1})
+// More information is available at:
+// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
+//
+// The (x,f) input data points must be sorted along x.
+// One can use github.com/gonum/stat.SortWeighted to do that.
+// The x and f slices must be of equal length and have length > 1.
+func Simpsons(x, f []float64) float64 {
+	validateInterval(x, f)
+
+	var integral float64
+
+	// If only three points are supplied,
+	// approximate the integral
+	// assuming the seccond coordinate points
+	// suffices as a midpoint and interpolate.
+	if len(x) == 3 {
+		return (x[2] - x[0]) * (f[0]/6 + 4/6*f[1] + 4*f[2]/6)
```

Comment:
`4/6` evaluates to 0, and the last term shouldn't have `4*`. These bugs are not catched by the tests because this line is not covered by them. It should. In addition, this again ignores the location of the middle sample, it should not.

Reply:
Thanks for catching this.

I'm abusing Newton-Cotes as a general rule here, but I feel like this equation isn't ignoring the midpoint, so to speak?

I've added a condition to ensure uniqueness between the points if only three are provided.

When performing Lagrangian interpolation for weight values to construct a polynomial with three points a=x[0], m=x[1], and b=x[2] where a < m <  b given:
```
L_i (x) =   f(a) * ((x-m) * (x-b)) / ((a-m) * (a-b))
         + f(m) * ((x-a) * (x-b)) / ((m-a) * (m-b))
         + f(b) * ((x-a) * (x-m)) / ((b-m)*(b-a))
```
i=1 is not valid. Additionally the endpoints in the three point form should produce a whole weight, to the best of my understanding. For example at i=0,1
```
// at L_1
integral += f[0] * ((x[0] - x[1]) * (x[0] - x[2]))
	         / ((x[0] - x[1]) * (x[0] - x[2]))
```
and
```
// at L_2
integral += f[1] * (x[2] - x[0]) * (x[2] - x[1])
	         / (x[2] - x[0]) * (x[2] - x[1])
```

I feel the form f[0]/6 + 4*f[1]/6 + f[2]  does account for the midpoint when evaluating the integral of a three point approximation with Simpson's method.

Reply:
It ignores `x[1]`. It assumes that the interpolating quadratic passes through `f[1]` located at `0.5*(x[0]+x[2])`.

Reply:
Without knowing the rate of change at all values of x and only having 3 points to work with, I think using the Simpson's method is most appropriate for the Simpson's formula.

`(x[2] - x[0]) * (f[0]/6.0 + 4.0*f[1]/6.0 + f[2]/6.0)`

We are only given three points and no derivative values. I think we run into the same issues with methods that don't assume equidistantly spaced points (e.g divided difference) due to the fact that evaluating the derived polynomial at the same points of x will cancel out like terms and evaluating the derived polynomial at an offset (e.g. 1e-4) introduces additional error.

```
var dividedDifferenceInterpolation = func(p float64) float64 {
            return f[0] +
                   (f[1]-f[0])/(x[1]-x[0]) +
                   (p-x[0])*(p-x[1])*((f[2]-f[1])/(x[2]-x[1])-(f[1]-f[0])/(x[1]-x[0]))/(x[2]-x[1])
        }

return (x[1]-x[0])*(dividedDifferenceInterpolation(x[1])-dividedDifferenceInterpolation(x[2]))*0.5 +
       (x[2]-x[1])*(dividedDifferenceInterpolation(x[2])-dividedDifferenceInterpolation(x[1]))*0.5
```

Reply:
The formula
```
(x[2] - x[0]) * (f[0]/6.0 + 4.0*f[1]/6.0 + f[2]/6.0)
```
ignores the value of `x[1]`. But why are you special-casing this at all? Why the general case cannot handle it?